ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/contract/CodeCache.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.ethereum.core.contract;
+
+import org.hyperledger.besu.ethereum.core.Account;
+import org.hyperledger.besu.ethereum.vm.Code;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheStats;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableMap;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class CodeCache implements LoadingCache<Account, Code> {
+
+  private final LoadingCache<Account, Code> cache;
```

Comment:
I would key by `Hash` instead of account.  Many accounts share code (such as most ERC-721s and ERC-20s) and this would reduce cache misses.  

Except for a convenience method (thunking account to account.getCodeHash) I would not use `Account` in this class, it's not a very reliable key.

Reply:
Switching to key by hash also opens up persisting the cache off-memory.

Reply:
Added a wrapper class to intercept hashCode() from guava. Was surprised to see there was no way to add a custom comparator to the cache via its builder.